dns-server/src/dns_server/mod.rs

```diff
@@ -211,15 +246,69 @@ impl AuthorityImpl {
 
         // AAAA records
         let mut ipv6_rec = RecordSet::with_ttl(self.host.clone(), RecordType::AAAA, TTL_IP);
-        for ip in ipv6 {
+        for ip in selected_ipv6_addrs {
             ipv6_rec.add_rdata(RData::AAAA(ip.into()));
         }
         new_records.insert(
             RrKey::new(ipv6_rec.name().clone().into(), ipv6_rec.record_type()),
             Arc::new(ipv6_rec),
         );
 
-        *self.inner.records_mut().await = new_records;
+        Some(new_records)
+    }
+
+    async fn refresh(&self) {
+        let new_records = self.create_records(&mut make_pseudo_rng());
+
+        if let Some(new_records) = new_records {
+            *self.inner.records_mut().await = new_records;
+        }
+    }
+
+    fn select_addresses<Addr: Clone>(
+        &self,
+        addrs: &BTreeMap<Addr, SoftwareInfo>,
+        count: usize,
+        rng: &mut impl Rng,
+    ) -> Vec<Addr> {
+        let same_software_info = SoftwareInfo::current(&self.chain_config);
+
+        let mut same_version_addrs = addrs
+            .iter()
+            .filter(|(_, software_info)| **software_info == same_software_info)
+            .map(|(addr, _)| addr.clone())
+            .choose_multiple(rng, count);
+        same_version_addrs.shuffle(rng);
+
+        let mut other_version_addrs = addrs
+            .iter()
+            .filter(|(_, software_info)| **software_info != same_software_info)
+            .map(|(addr, _)| addr.clone())
+            .choose_multiple(rng, count);
+        other_version_addrs.shuffle(rng);
+
+        #[allow(clippy::float_arithmetic)]
+        let same_version_addrs_preferred_count =
+            (count as f64 * SAME_SOFTWARE_VERSION_PEERS_RATIO) as usize;
+
+        let mut result = Vec::with_capacity(count);
+
+        // First take the required number of same-version addresses.
```

Comment:
I think there's one corner case here that has to be covered with much more care. Before partitioning the software, we have to check the available ratio of new/(old+new) versions. If it's already heigher than `SAME_SOFTWARE_VERSION_PEERS_RATIO`, then nothing needs to be done, and a random selection of the available set is enough. This is because we do not want to increase the probability of picking an old node, which may be a side effect of what we're doing here, but we want to cap it at `1 - SAME_SOFTWARE_VERSION_PEERS_RATIO`.